src/autofix/autofix.ts

```diff
@@ -0,0 +1,318 @@
+	skipLibCheck: true,
+
+	target: ts.ScriptTarget.ES2022,
+	module: ts.ModuleKind.ES2022,
+	isolatedModules: true,
+	sourceMap: true,
+	suppressOutputPathCheck: true,
+	noLib: true,
+	noResolve: true,
+	allowNonTsExtensions: true,
+};
+
+function createProgram(inputFileNames: string[], host: ts.CompilerHost): ts.Program {
+	return ts.createProgram(inputFileNames, compilerOptions, host);
+}
+
+// eslint-disable-next-line @typescript-eslint/require-await
+export default async function ({
+	rootDir: _unused1,
+	namespace: _unused2,
+	resources,
+	context,
+}: AutofixOptions): Promise<AutofixResult> {
+	const sourceFiles: SourceFiles = new Map();
+	const resourcePaths = [];
+	for (const [resourcePath, resource] of resources) {
+		const sourceFile = ts.createSourceFile(
+			resourcePath,
+			resource.content,
+			{
+				languageVersion: ts.ScriptTarget.ES2022,
+				jsDocParsingMode: ts.JSDocParsingMode.ParseNone,
+			}
+		);
+		sourceFiles.set(resourcePath, sourceFile);
+		resourcePaths.push(resourcePath);
+	}
+
+	const compilerHost = createCompilerHost(sourceFiles);
+	const program = createProgram(resourcePaths, compilerHost);
+
+	const checker = program.getTypeChecker();
+	const res: AutofixResult = new Map();
+	for (const [resourcePath, sourceFile] of sourceFiles) {
+		const newContent = applyFixes(checker, sourceFile, resourcePath, resources.get(resourcePath)!, context);
+		if (newContent) {
+			res.set(resourcePath, newContent);
+		}
+	}
+
+	return res;
+}
+
+function applyFixes(
+	checker: ts.TypeChecker, sourceFile: ts.SourceFile, resourcePath: ResourcePath,
+	resource: AutofixResource, context: LinterContext
+): string | undefined {
+	const {content} = resource;
+
+	// Group messages by id
+	const messagesById = new Map<MESSAGE, RawLintMessage[]>();
+	for (const msg of resource.messages) {
+		if (!messagesById.has(msg.id)) {
+			messagesById.set(msg.id, []);
+		}
+		messagesById.get(msg.id)!.push(msg);
+	}
+
+	const changeSet: ChangeSet[] = [];
+	let existingModuleDeclarations = new Map<ts.CallExpression, ExistingModuleDeclarationInfo>();
+	if (messagesById.has(MESSAGE.NO_GLOBALS)) {
+		existingModuleDeclarations = generateSolutionNoGlobals(
+			checker, sourceFile, content,
+			messagesById.get(MESSAGE.NO_GLOBALS) as RawLintMessage<MESSAGE.NO_GLOBALS>[],
+			changeSet, []);
+	}
+
+	for (const [defineCall, moduleDeclarationInfo] of existingModuleDeclarations) {
+		addDependencies(defineCall, moduleDeclarationInfo, changeSet, context.getMetadata(resourcePath));
+	}
+
+	if (changeSet.length === 0) {
+		// No modifications needed
+		return undefined;
+	}
+	return applyChanges(content, changeSet);
+}
+
+function addDependencies(
+	defineCall: ts.CallExpression, moduleDeclarationInfo: ExistingModuleDeclarationInfo,
+	changeSet: ChangeSet[], resourceMetadata: LintMetadata
+) {
+	const {moduleDeclaration, importRequests} = moduleDeclarationInfo;
+
+	const existingImportModules = resourceMetadata.transformedImports.get("sap.ui.define") ?? [];
```

Comment:
There is no guarantee that `transformedImports` is filled. If the transforming failed for whatever reason, we do not have this information available.

Reply:
JS files would always go through the transformer, therefore this would be set. TypeScript would have catched this one.
Only TS files would fail, but we do not support them for now either ways.

I'll add a guard about it

Reply:
The transformation for JS files *can* fail in certain cases (currently for example when de-structuring is used within factory arguments) and the error would be caught, so the file remains as it is.

Reply:
Can you provide an example?
I'm currently trying with the following and it's parsed successfully:

```js
sap.ui.define(["sap/m/Button", "sap/m/library"], function(ButtonRenamed, {BackgroundDesign}) {
```

Reply:
I sent you the file via chat

Reply:
Thanks!